include/zephyr/sys/async.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Public APIs for Async utilities
+ *
+ * Functions and datatypes useful for safely communicating to a thread or work
+ * queue of an asynchronous functions completion.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ASYNC_H_
+#define ZEPHYR_INCLUDE_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**
+ * @brief Async operation callback type
+ *
+ * Common function pointer type used for asynchronous completion callbacks
+ * and implemented for some commonly used IPC notification mechanisms.
+ *
+ * May also be implemented by an application with the understanding
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
```

Comment:
As discussed earlier in the API meeting, I think the common async callback function signature needs to be able to take a context. For device drivers, the context will be the `const struct device *` for the device associated with the notification.

```suggestion
typedef void (*async_callback_t)(void *context, int result, void *userdata);
```

Reply:
Perhaps the `struct async_*`  structures could include a `void *context` pointer as well?

Reply:
Where would that be used? An example would be good to help me understand the use case

Reply:
It would be used by the consumer of the async event, similar as to where `result` would be used.